unity-environment/Assets/ML-Agents/Scripts/Agent.cs

```diff
@@ -926,37 +937,42 @@ void MakeRequests(int academyStepCounter)
         /// Converts a camera and correspinding resolution to a 2D texture.
         /// </summary>
         /// <returns>The 2D texture.</returns>
-        /// <param name="camera">Camera.</param>
+        /// <param name="obsCamera">Camera.</param>
         /// <param name="width">Width of resulting 2D texture.</param>
         /// <param name="height">Height of resulting 2D texture.</param>
-        public static Texture2D ObservationToTexture(Camera camera, int width, int height)
+        /// <param name="texture2D">Texture2D to render to.</param>
+        public static Texture2D ObservationToTexture(Camera obsCamera, int width, int height, Texture2D texture2D)
         {
-            Rect oldRec = camera.rect;
-            camera.rect = new Rect(0f, 0f, 1f, 1f);
+            Rect oldRec = obsCamera.rect;
+            obsCamera.rect = new Rect(0f, 0f, 1f, 1f);
             var depth = 24;
             var format = RenderTextureFormat.Default;
             var readWrite = RenderTextureReadWrite.Default;
 
             var tempRT =
                 RenderTexture.GetTemporary(width, height, depth, format, readWrite);
-            var tex = new Texture2D(width, height, TextureFormat.RGB24, false);
+            
+            if (width != texture2D.width || height != texture2D.height)
+            {
+                texture2D = new Texture2D(width, height, TextureFormat.RGB24, false);
+            }
 
             var prevActiveRT = RenderTexture.active;
-            var prevCameraRT = camera.targetTexture;
+            var prevCameraRT = obsCamera.targetTexture;
 
             // render to offscreen texture (readonly from CPU side)
             RenderTexture.active = tempRT;
-            camera.targetTexture = tempRT;
+            obsCamera.targetTexture = tempRT;
 
-            camera.Render();
+            obsCamera.Render();
 
-            tex.ReadPixels(new Rect(0, 0, tex.width, tex.height), 0, 0);
-            tex.Apply();
-            camera.targetTexture = prevCameraRT;
-            camera.rect = oldRec;
+            texture2D.ReadPixels(new Rect(0, 0, texture2D.width, texture2D.height), 0, 0);
+            texture2D.Apply();
+            obsCamera.targetTexture = prevCameraRT;
+            obsCamera.rect = oldRec;
             RenderTexture.active = prevActiveRT;
             RenderTexture.ReleaseTemporary(tempRT);
-            return tex;
+            return texture2D;
```

Comment:
Can you just pass texture 2D by reference? Use `out Texture2D texture2D` as method parameter.

Reply:
`out` assumes you will initialize within the function, which is explicitly what we don't want. I replaced it with a `ref` instead though.